libcontainer/env.go

```diff
@@ -0,0 +1,56 @@
+package libcontainer
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+)
+
+// prepareEnv checks supplied environment variables for validity, removes
+// duplicates (leaving the last value only), and sets PATH from env, if found.
+// Returns the deduplicated environment, and a flag telling if HOME is found.
+func prepareEnv(env []string) ([]string, bool, error) {
+	// Clear the current environment (better be safe than sorry).
+	os.Clearenv()
+
+	if env == nil {
+		return nil, false, nil
+	}
+	// Deduplication code based on dedupEnv from Go 1.22 os/exec.
+
+	// Construct the output in reverse order, to preserve the
+	// last occurrence of each key.
+	out := make([]string, 0, len(env))
+	saw := make(map[string]bool, len(env))
+	for n := len(env); n > 0; n-- {
+		kv := env[n-1]
+		i := strings.IndexByte(kv, '=')
+		if i == -1 {
+			return nil, false, errors.New("invalid environment variable: missing '='")
+		}
+		if i == 0 {
+			return nil, false, errors.New("invalid environment variable: name cannot be empty")
+		}
+		key := kv[:i]
+		if saw[key] { // Duplicate.
+			continue
+		}
+		saw[key] = true
+		if strings.IndexByte(kv, 0) >= 0 {
+			return nil, false, fmt.Errorf("invalid environment variable %q: contains nul byte (\\x00)", key)
+		}
+		if key == "PATH" {
+			// Needs to be set as it is used for binary lookup.
+			if err := os.Setenv("PATH", kv[5:]); err != nil {
```

Comment:
I'm unsure about doiong this here vs in the system.Exec() func. The latter seems more tidy, IMHO. I guess it is only needed there, right? Or maybe some hooks that exec before the container need it too?

Reply:
> len(key) +1? Or define val here using that and assign val?

Changed to `i+1` (might result in worse asm as it's not constant but who cares).

> I'm unsure about doiong this here vs in the system.Exec() func.

For both `runc start/run` and `runc init` we do explicit binary lookup (`exec.LookPath`) slightly earlier while we still have a chance to return an error (as we close everything before doing exec, we won't be able to properly return an error), and when we use a looked up binary path.

Now, if there was an `exec.LookPath` version which allows to specify PATH, we could use it instead of setting PATH in the current context. Alas, it doesn't exist and I don't want to replicate it in runc.

NOTE that currently (i.e. before this PR) we set *all* the environment variables from spec in the runc parent context (which can be bad in case someone would use GODEBUG or GOEXPERIMENT etc). This PR reduces *all* to PATH.

Reply:
Makes sense, thanks for the explanation!